src/NuGet.Core/NuGet.Common/KeyedMutex.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace NuGet.Common
+{
+    internal sealed class KeyedMutex : IDisposable
+    {
+        private Dictionary<string, LockState> _locks;
+        private SemaphoreSlim _dictionaryLock;
+
+        internal KeyedMutex()
+        {
+            _locks = new Dictionary<string, LockState>();
+            _dictionaryLock = new SemaphoreSlim(initialCount: 1);
+        }
+
+        internal async Task EnterAsync(string key, CancellationToken token)
```

Comment:
For each non-private method with a `key` parameter, verify the argument is not null.

```C#
if (key == null)
{
    throw new ArgumentNullException(nameof(key));
}
```